### PR TITLE
Add REST documentation and automatically regenerate and deploy documentation on updates

### PR DESCRIPTION
**Pull request by wpaulino**

### CODE REVIEW

**Comment by halseth** [comment]
This is awesome! Is there a way to see to look at the documentation generated?

File: source/index.html.md [hunk 1]
```diff
@@ -11,36 +11,40 @@ toc_footers:
...
+used for RPC authentication. The examples to the right will show how these can
+be used in order to make a successful, secure, and authenticated gRPC request.
 
-Lastly, keep in mind that the code examples will differ slightly based on your
-operating system and specific setup. The `LND_HOMEDIR` used in the gRPC examples
-is `~/.lnd/` for Linux or `~/Library/Application Support/Lnd/tls.cert` for Mac
+The original `rpc.proto` file from which the gRPC documentation was generated
+can be found [here](https://github.com/lightningnetwork/lnd/blob/master/lnrpc/rpc.proto).
 
+Alternatively, the REST documnetation can be found [here](https://api.lightning.community/rest).
```
Comment (halseth) [inline source/index.html.md]: should use a relative link

File: update_and_render.sh [hunk 2]
```diff
@@ -5,7 +5,15 @@ curl -o rpc.proto -s https://raw.githubusercontent.com/lightningnetwork/lnd/mast
 # Generate the rpc.json file from rpc.proto, so that generate_slate_docs.py can parse it
 protoc -I. -I$GOPATH/src/github.com/grpc-ecosystem/grpc-gateway/third_party/googleapis --doc_out=json,rpc.json:. rpc.proto
 
-# Generate the new docs
+# Update lncli to the respective commit.
+pushd $GOPATH/src/github.com/lightningnetwork/lnd
+git pull
+git reset --hard $commit
```
Comment (Roasbeef) [inline update_and_render.sh]: Why not a `git checkout`?

File: grpc.py [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+    Example output:
+        ('getinfo', "GetInfo serves a request. It returns info.")
+    """
+
+    if description is None:
+        return None, None
+
+    # Match lncli: `lncli_name` or lncli:`lncli_name`.
+    command = None
+    match = re.search(r'(?:lncli:.?`)(.*)(?:`)', description)
```
Comment (Roasbeef) [inline grpc.py]: such regex, wow

File: grpc.py [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+
+        if param['label'] == 'repeated':
+            p['type'] = 'array ' + param['type']
+        else:
+            p['type'] = param['type']
+
+        # If the type includes an "lnrpc." prefix, then we'll remove it in order
+        # to include a link to the message being referenced.
+        lnrpc_prefix = "lnrpc."
+        if param['fullType'][:len(lnrpc_prefix)] == lnrpc_prefix:
```
Comment (Roasbeef) [inline grpc.py]: Not a blocker, but could instead use `.startswith(..)` here.

File: rest.py [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+
+
+def parse_rest_definition_name(name):
+    """
+    Convenience method that removes the "lnrpc" prefix from the name of a REST
+    definition.
+    """
+
+    prefix = "lnrpc"
+    if prefix != name[:len(prefix)]:
```
Comment (Roasbeef) [inline rest.py]: Similar comment here w.r.t using `startswith`.

File: server.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+def update_and_render(commit_sha):
+    call('./update_and_render.sh {}'.format(commit_sha), shell=True)
+
+
+def deploy_updated_docs():
+    call('./deploy.sh', shell=True)
+
+
+@app.route('/update', methods=['POST'])
+def handle_proto_update():
```
Comment (Roasbeef) [inline server.py]: niiiice

File: README.md [hunk 7]
```diff
@@ -26,6 +26,7 @@ You're going to need:
 
  - **Linux or OS X** — Windows may work, but is unsupported.
  - **Ruby, version 2.2.5 or newer**
+ - **Python, version 3 or newer**
```
Comment (Roasbeef) [inline README.md]: Your ancestors salute you 💂

File: update_and_render.sh (hunk 2 above)
Comment (Roasbeef) [reply update_and_render.sh]: Tried this locally just now, and had to manually approve a merge in vim. Will need to ensure that the update bot doesn't get stuck on something similar.

File: update_and_render.sh [hunk 8]
```diff
@@ -1,13 +1,24 @@
...
+
+# Generate the rpc.json file from rpc.proto, so that render.py can parse it.
 protoc -I. -I$GOPATH/src/github.com/grpc-ecosystem/grpc-gateway/third_party/googleapis --doc_out=json,rpc.json:. rpc.proto
 
-# Generate the new docs
+# Update lncli to the respective commit.
+pushd $GOPATH/src/github.com/lightningnetwork/lnd
+git pull
+git reset --hard $commit
+make
```
Comment (Roasbeef) [inline update_and_render.sh]: We don't need to make anything. We only care about the protos.

**Comment by Roasbeef** [comment]
REST looks fine locally, but when I render, I get
screen shot 2018-06-13 at 8 05 30 pm

File: source/index.html.md (hunk 1 above)
Comment (wpaulino) [reply source/index.html.md]: Fixed.

File: update_and_render.sh (hunk 2 above)
Comment (wpaulino) [reply update_and_render.sh]: Fixed.

File: grpc.py (hunk 4 above)
Comment (wpaulino) [reply grpc.py]: Fixed.

File: rest.py (hunk 5 above)
Comment (wpaulino) [reply rest.py]: Fixed.

File: update_and_render.sh (hunk 8 above)
Comment (wpaulino) [reply update_and_render.sh]: We'll need it in order to provide updated documentation for the `lncli` shell commands.

**Pull request merged**